### PR TITLE
Add Beacon Monitor

### PR DESCRIPTION
**Pull request by Adi146**
## Summary

This Pull Request adds a sensor for monitoring iBeacon. It will add a sensor which shows the distance to the closest beacon. There are all beacons in range listed in the attributes.

## Screenshots

Screenshot_20220424-204800
Screenshot_20220424-204806

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#736

## Any other notes

Issue: https://github.com/home-assistant/android/issues/900

### CODE REVIEW

File: common/src/main/res/values/strings.xml
Comment (dshokouhi) [inline common/src/main/res/values/strings.xml:430]: We like to use the description as a place to describe the various settings a sensor could have. Could you write up about each setting offered making sure to mention units if applicable? Also if this feature will have a big impact on battery its also good to mention that here so users are aware of the implications.

Comment (dshokouhi) [inline common/src/main/res/values/strings.xml:430]: We like to use the description as a place to describe the various settings a sensor could have. Could you write up about each setting offered making sure to mention units if applicable? Also if this feature will have a big impact on battery its also good to mention that here so users are aware of the implications.

Comment (Adi146) [reply common/src/main/res/values/strings.xml]: I updated the description, let me know if something is not explained good enough or I missed something

**Comment by dshokouhi** [comment]
Started taking a look at the debug APK today.  Just a couple of questions so far based on usability.

- What should the state of the sensor be when it does not detect any beacons within range? Like if we enable the sensor and are not near any beacons.
- Should the state of the sensor reflect the state of the monitor (setting) being enabled? The reason I say this is because users will want to know if the monitor is enabled when they want to control it via notification command. We do something similar for the beacon transmitter today. I think maybe combined with the first question we can probably introduce states like `not monitoring` or `monitoring` when the state is not the closest distance?

**Comment by Adi146** [comment]
At the moment the sensor shows 0.0m if there is no beacon nearby. If the monitor is turned off it simply stops updating and keeps the last state. I agree, we should probably add some indication if it is monitoring or not but I don't think we should do that with the state because this would break the unit. What do you think about adding another attribute, like `state: enabled` or `state: disabled`?

**Comment by dshokouhi** [comment]
> At the moment the sensor shows 0.0m if there is no beacon nearby.

About how long would you say it would take for this to show up? 2 minutes? If I enable the sensor after a fresh install and I do not have any beacons nearby the state of the sensor shows as enabled and I never see the sensor being registered or data being sent to HA for it.  Looking at the code it looks like it only sends an update if it has found a beacon before or if one expired but not if it didn't find anything at all on the first try.  Is that correct or did I misinterpret when we send updates?



> I agree, we should probably add some indication if it is monitoring or not but I don't think we should do that with the state because this would break the unit. What do you think about adding another attribute, like `state: enabled` or `state: disabled`?

I think that is fine as it still describes the state of the sensor, so users should expect the state to be 0.0m if monitoring is disabled correct? 

One more question :) Do we also need to account for bluetooth being turned off? Or is that a none issue? I guess I need to find a beacon to test this out better 😄 I ask because in the beacon transmitter sensor we do need to account for it.

**Comment by Adi146** [comment]
> About how long would you say it would take for this to show up? 2 minutes? If I enable the sensor after a fresh install and I do not have any beacons nearby the state of the sensor shows as enabled and I never see the sensor being registered or data being sent to HA for it. Looking at the code it looks like it only sends an update if it has found a beacon before or if one expired but not if it didn't find anything at all on the first try. Is that correct or did I misinterpret when we send updates?

You got the sensor updates correctly and thats a case I didn't consider. I will have to add a sensor update on enable/disable for the state attribute anyways so this should fix it.

> I think that is fine as it still describes the state of the sensor, so users should expect the state to be 0.0m if monitoring is disabled correct?

Yes, if the sensor is disabled the state will be 0.0m and attribute `state` will be `disabled`. If there is no beacon in range the state will also be 0.0m but the attribute `state`will be `enabled`

> One more question :) Do we also need to account for bluetooth being turned off? Or is that a none issue?

Bluetooth should be enabled automatically when a scan period starts so this should be no big deal.

I hope, I can implement these changes later today.

**Comment by dshokouhi** [comment]
> Yes, if the sensor is disabled the state will be 0.0m and attribute `state` will be `disabled`. If there is no beacon in range the state will also be 0.0m but the attribute `state`will be `enabled`

So in this particular case where the state value has not changed, the attribute will not update either. We only update sensors if the state has changed since the last update as attributes are supposed to describe the state of the sensor so they should not update when the state has not changed. Maybe its better if the state is just `0` when the monitor is disabled and `0.0` when its enabled so users can see it change if they enable it and are not near any beacons?

**Comment by dshokouhi** [comment]
is there a beacon app I can use on another device to test with? I tried to use the apps transmitter sensor on a fire tablet but it doesnt seem to be picking it up on my Pixel in the same room but I am also not sure if its designed for that type of beacon :) I was thinking if we can use the apps own sensors it might be worth a write up to show it working between 2 devices using the app

**Comment by madface303** [comment]
On android you can use beacon scope to send ibeacons

**Comment by Adi146** [comment]
> > Yes, if the sensor is disabled the state will be 0.0m and attribute `state` will be `disabled`. If there is no beacon in range the state will also be 0.0m but the attribute `state`will be `enabled`
> 
> So in this particular case where the state value has not changed, the attribute will not update either. We only update sensors if the state has changed since the last update as attributes are supposed to describe the state of the sensor so they should not update when the state has not changed. Maybe its better if the state is just `0` when the monitor is disabled and `0.0` when its enabled so users can see it change if they enable it and are not near any beacons?

After all it seems like we need to account for if Bluetooth is on or off. What do you think of adding a second sensor which shows the state of the monitor (enabled, disabled, bluetooth off) like with it is implemented in the Beacon Transmitter?

**Comment by madface303** [comment]
Just a question, is there anything special to do to detect a beacon? I have an esp32 beacon which is recognized by other scanners, but the debug apk i installed beside the normal app doesn't find it. The sensor sais monitoring but there are no attributes.
In the log i see an error about the permission, to test i gave the debug app every permission possible but the error persists.

```
04-29 20:46:47.258 23562 28427 I BluetoothAdapter: STATE_ON
04-29 20:46:47.258 23562 28427 D BluetoothLeScanner: could not find callback wrapper
04-29 20:46:48.255 23562 28427 I BluetoothAdapter: STATE_ON
04-29 20:46:48.255 23562 28427 D BluetoothLeScanner: could not find callback wrapper
04-29 20:46:48.258 23562 28427 I BluetoothAdapter: STATE_ON
04-29 20:46:48.259 23562 28427 I BluetoothAdapter: STATE_ON
04-29 20:46:48.259 23562 28427 I BluetoothAdapter: STATE_ON
04-29 20:46:48.260 23562 28427 D BluetoothLeScanner: Start Scan with callback
04-29 20:46:48.265 23562 28427 E CycledLeScannerForLollipop: Cannot start scan.  Security Exception: Need android.permission.BLUETOOTH_SCAN permission for AttributionSource { uid = 10393, packageName = io.homeassistant.companion.android.debug, attributionTag = null, token = android.os.BinderProxy@3bdb84a, next = null }: GattService registerScanner
```

I am running on a galaxy s21+ with samsung stock android 12

**Comment by dshokouhi** [comment]
hmm I don't believe I noticed that permission error before but its possible that is why the BLE transmitter does not get picked up by other devices emitting it as I am on android 12 as well.

```
04-29 20:46:48.265 23562 28427 E CycledLeScannerForLollipop: Cannot start scan.  Security Exception: Need android.permission.BLUETOOTH_SCAN permission for AttributionSource { uid = 10393, packageName = io.homeassistant.companion.android.debug, attributionTag = null, token = android.os.BinderProxy@3bdb84a, next = null }: GattService registerScanner
```

This is a new permission in API level 31 so we probably just need to add it to our permission check as it is missing there: https://developer.android.com/reference/android/Manifest.permission#BLUETOOTH_SCAN

**Comment by Adi146** [comment]
I added the permission, can you check again?

**Comment by madface303** [comment]
Tried, but with this actual "action build" the sensor cannon be turned on. Nothing to see in the logs about the beacon scanner

**Comment by Adi146** [comment]
> Tried, but with this actual "action build" the sensor cannon be turned on. Nothing to see in the logs about the beacon scanner

Thanks for the feedback. It should be fixed now, at least I could enable it with the emulator. I'm still using Android 11 on my phone so its a bit hard to test.

**Comment by dshokouhi** [comment]
Working for me during initial testing. Good news too is that it is picking up the BLE transmitter beacon from the android app installed on other devices (wall mounted fire tablet 😃 ).

Great job!!  Will run some more tests and start looking at the code too.  

image

Lets make sure to update the docs to match the new state logic.

**Comment by madface303** [comment]
Same for me, initial test is fine. Will do more practise tests tomorrow, but the permission problem seems to be solved!

**Comment by madface303** [comment]
A short question how it is meant to be integrated into ha. I now have the sensor with 4 beacon attributes (only one of them is mine), i can easily make a template sensor of the attribute of my beacon, but when i leave this stays at the last distance submitted. From my opinion as a user it would be fine if we can a) filter our own beacons out directly in the app so that only my beacons are submitted to ha and b) the state of the sensor should be the distance again (and 0 if no beacon is in range) and the state of the monitor should go to the attributes. So even a user with a not so large experience (as me for example ;) ) can easily build an automation for being home/not_home. What are you thinking about?

**Comment by Adi146** [comment]
> A short question how it is meant to be integrated into ha. I now have the sensor with 4 beacon attributes (only one of them is mine), i can easily make a template sensor of the attribute of my beacon, but when i leave this stays at the last distance submitted. From my opinion as a user it would be fine if we can

Here is my template sensor how I'm using it at the moment.
```
{% set beacon_uuid = states.sensor.smartphone_test_beacon_monitor.attributes | items | selectattr('0', 'in', '[bc59aec6-38b0-a4e9-58ce-ff107994091a,afe0fbaa-755e-64b5-1740-71570fa08335 2cc98e32-6ea7-e417-fb2f-baaa237ea789]') | sort(attribute=1) | first | first %}
{% if beacon_uuid == 'bc59aec6-38b0-a4e9-58ce-ff107994091a' or beacon_uuid == 'afe0fbaa-755e-64b5-1740-71570fa08335' %}
   Room1
{% elif beacon_uuid == '2cc98e32-6ea7-e417-fb2f-baaa237ea789' %}
  Bathroom
{% else %}
   Away
{% endif %}
```

I have three beacons of which two are in a bit larger room and one is in the bathroom. The template first filters for my beacons and then sorts them by the distance. After that it checks which beacon is the closest and reports the room I'm in.

>a) filter our own beacons out directly in the app so that only my beacons are submitted to ha

There are a lot of features which could be implemented later but for now I just want to get a basic sensor working. 

>b) the state of the sensor should be the distance again (and 0 if no beacon is in range) and the state of the monitor should go to the attributes. So even a user with a not so large experience (as me for example ;) ) can easily build an automation for being home/not_home. What are you thinking about?

I think there are a lot of use cases where you just want to check if a specific beacon is in range and in this case the state implementation as I had it before would be pretty useless. With the current implementation you could use a template like this

```
{% set distance = state_attr('sensor.smartphone_test_beacon_monitor', 'bc59aec6-38b0-a4e9-58ce-ff107994091a') %}
{% if distance != None %}
  in range
{% else %}
  out of range
{% endif %}
```

**Comment by madface303** [comment]
This template would have taken me hours ;). Thanks a lot, for now i will create your sensor and do a long test run with it. But for my first experience it works every time i had a look at it, really good job. Another tool in the swif knife of home assistant companion!

File: app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt
Comment (dshokouhi) [inline app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt:334]: So I noticed that on a fresh install on android 12 if I dont enable the BT state sensor the state of this sensor remains at "Bluetooth is turned off"  eventhough BT is on for this device. As soon as I enable BT state sensor the monitor starts reporting beacons and the state updates too. I think we may need to change the check on `isBtOn()` to not be sensor specific like it is today. The BLE transmit sensor does not have this particular issue. Or we may just need to add more BT permissions for Android 12 devices to work around it.

Comment (Adi146) [reply app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt]: I changed it to the `BluetoothUtils.isOn(context)` function which seem to fix this case. However when I switch Bluetooth of in the emulator on Android 12 the sensor state keeps showing `Monitoring`. I'm not sure if this happens on real devices too. Could you please check this?

Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt]: latest debug APK works well on my pixel 6 pro :)  the state also updates immediately when I toggle BT on/off

**Comment by Adi146** [comment]
@dshokouhi whats the state of the review? Are we ready to merge?

**Review by JBassett — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/homeassistant/companion/android/bluetooth/ble/MonitoringManager.kt
Comment (Adi146) [reply app/src/main/java/io/homeassistant/companion/android/bluetooth/ble/MonitoringManager.kt:48]: I'm not sure if this works because `observeForever`  has to run in the main thread. Correct me if I'm wrong.

File: common/src/main/java/io/homeassistant/companion/android/database/sensor/SensorDao.kt
Comment (Adi146) [reply common/src/main/java/io/homeassistant/companion/android/database/sensor/SensorDao.kt:76]: We have to reset the `last_sent_state` before sending an sensor update, so that the attributes get updated in home-assistant.

Comment (JBassett) [reply common/src/main/java/io/homeassistant/companion/android/database/sensor/SensorDao.kt:76]: suspend functions are still run synchronously, they just don't block the thread they are on when waiting on I/O

File: app/src/main/java/io/homeassistant/companion/android/bluetooth/ble/MonitoringManager.kt
Comment (JBassett) [reply app/src/main/java/io/homeassistant/companion/android/bluetooth/ble/MonitoringManager.kt:48]: You can still scope a coroutine that uses the main thread.

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt
Comment (dshokouhi) [inline app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt:371]: Now that #1759 has been merged we will need to update this PR to switch from `TITLE` to `COMMAND`, we shoudl also update the documentation to reflect this change as well.

**Review by dshokouhi — CHANGES_REQUESTED** [review]
We need to adjust the notification commands now that the breaking change PR was merged. Asides from this there is also the failing linter that needs to be resolved.

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt
Comment (dshokouhi) [inline app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt:383]: We also need to switch to `jsonData[COMMAND]` here as well, just like the other commands above/below :)

**Review by dshokouhi — CHANGES_REQUESTED** [review]
The notification command was only partially updated :)

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt
Comment (Adi146) [reply app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt]: ups I overlooked that

**Review by dshokouhi — APPROVED** [review]
I have tested the debug APK and its working as expected in my tests. Lets wait for another review to make sure the previous requested changes are also as expected.

**Comment by dshokouhi** [comment]
@Adi146 just a heads up you have a conflict with a file here

File: app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt
Comment (dshokouhi) [inline app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt:345]: We should make sure none of the code below is reachable if the sensor is not enabled. May also want to do something similar with `updateBeaconMonitoringDevice()` up above. We do a similar check with other sensors so the code only runs if the sensor is enabled.

**Review by dshokouhi — CHANGES_REQUESTED** [review]
Apologize for not catching this earlier, just a small change to make sure the code is only reachable if the sensor is not enabled.

File: app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/sensors/BluetoothSensorManager.kt:360]: This function was updated in #2734 (which has been merged), please use `null` instead of `""` to force an update of the sensor

**Comment by Yogranov** [comment]
Hey,
Im here to say i wanted this feature for a long time (literally started an background app project for temp workaround).
Ive downloaded this artifact APK, and it looks amazing! :D
It was little hard to make an automation with ranges (less than 10m), but it must merge to next version guys :P
Thanks @Adi146 !

**Comment by Adi146** [comment]
I added the requested changes. It would be nice if we could finally merge this.

**Review by dshokouhi — APPROVED** [review]
All my requested changes have been met. We are planning on cutting a new release with existing beta features so this PR (given final review from others) will probably be included in the next beta cycle so it can be given a proper test cycle. We actually have other PRs we are holding off on as well.

**Comment by nodnarb402** [comment]
I joined the beta track specifically for this feature. Thanks for working on this. Very much looking forward to it getting merged!

**Pull request merged**
